Custom RefreshIndicator 생성

thomaskim1130 opened the pull request:
## Overview
<!-- 간단하게 이 PR이 무엇인지 설명해주세요. 예: 새로운 로그인 기능 추가 -->
로딩을 위해 pull down하는 거리 감소. 편의성 증진
## Changes
<!-- 이 PR로 인해 변경되는 사항을 나열해주세요. 예:
- 로그인 페이지 UI 업데이트
- 로그인 API 연동
- 로그인 에러 처리 추가
-->
- refreshIndicator.adaptive()를 ScrollNotification 클래스를 활용해 custom화 함. 
- 기존 pull down 50(변경 불가) 에서 30으로 조정.
- main_page.dart, post_list_show_page.dart, user_page.dart, post_view_page 등 페이지에 적용
## Implementaion Method
<!-- 변경사항을 구현한 방법에 대해 특별히 전달해야할 것이 있다면 설명해주세요. 예:
- React Hook을 사용하여 상태 관리
- JWT를 사용한 인증 처리
-->
- global_key.dart에 refreshIndicatorKey 추가 (페이지별로 개별 생성)
- utils/refresh_indicator.dart를 추가하여 메소드 customRefreshIndicator 생성
## After Changes
<!-- 변경 사항을 확인 할 수 있는 방법을 알려주세요. 가능하다면, 변화를 보여주는 스크린샷 또는 동영상을 첨부해주세요. 예:
- 로그인 성공 시 홈페이지로 리다이렉트하는 동영상
-->
- scrollDownLength를 customize할 수 있음
## Related Issues
<!-- 관련 버그 현상이나 관련 이슈 번호가 있다면 링크를 적어주거나 스크린샷을 첨부해주세요. 예: #123 -->

## Rollback Scenario
<!-- 해당 PR를 롤백하는 방법과 순서를 알려주세요. 예: 
1. revert commit
2. .env 파일 수정
-->
1. customRefreshIndicator 대신 refreshIndicator.adaptive 사용
2. globalKey 모두 제거
3. AlwaysScrollablePhysics의 parent 변경하기
## TODO
<!-- 앞으로 해야하는 것이나, 논의가 필요한 부분 등을 적어주세요. 예:
- 로그인 관련 테스트 케이스 추가
- 다국어 지원 검토
-->
- user_view_page.dart 에도 적용. (DONE)
- refreshIndicator 다수 중복. Wrapper 함수 제작? (DONE)

thomaskim1130 committed:
Feat(global_key.dart): Custom RefreshIndicator 생성

refreshIndicatorKey (gloabl_key.dart 내부에도 저장) 를 생성해 trigger를 위해 pull down해야 하는 거리를 조정함 (50 -> 20).

위를 main_page, user_page, post_list_show_page 에만 적용함.

thomaskim1130 committed:
Fix(main_page.dart): global key duplicate

GlobalKey 'refreshIndicatorKey'를 페이지별로 생성하여 duplicate global key 이슈를 방지
추후 다른 파일에도 유사하게 적용할 예정

thomaskim1130 committed:
Fix: Android Notification Fix

Android에서 ScrollNotification이 negative일 때 (즉 위에서 아래로 outOfRange로 스크롤 할 때) 의 pixel을 잡지 못하는 상황을 개선함.
구체적으로, parent: BouncingScrollPhysics()를 AlwaysScrollableScrollPhysics에 넣어 android도 ios와 동일한 설정으로 바꿈.

thomaskim1130 committed:
Add(refresh_indicator.dart): created customRefreshIndicator

기존의 RefreshIndicator.adaptive를 계승하는 customRefreshIndicator 메소드를 정의함 (utils/refresh_indicator.dart).
Wrapper 함수 형식으로 생성하여 공식 문서상 parameter를 모두 가져옴.

thomaskim1130 committed:
Refactor: applied customRefreshIndicator

customRefreshIndicator를 main_page, notification_page, post_list_show_page, post_view_page, user_page, user_view_page 에 모두 적용함.

sangohkim commented on lib/pages/post_view_page.dart:
```diff
@@ -277,8 +280,8 @@ class _PostViewPageState extends State<PostViewPage> {
-                                  physics:
-                                      const AlwaysScrollableScrollPhysics(),
+                                  physics: const AlwaysScrollableScrollPhysics(
+                                      parent: BouncingScrollPhysics()),
```
원래 안드로이드 기종에서는 바운싱 효과가 없는게 기본설정이어서 parent: BouncingScrollPhysics()를 설정해주게 되면 바운싱 효과가 나타나게됩니다. 문제가 될 부분은 전혀 없는데 혹시 의도하신 부분인지가 궁금합니다!

sangohkim commented on lib/utils/refresh_indicator.dart:
```diff
@@ -0,0 +1,71 @@
+import 'package:flutter/material.dart';
```
위젯을 생성하는 기능이니 utils 보다는 widgets 디렉터리에 위치시키는게 좋을것 같습니다.

sangohkim commented on lib/utils/global_key.dart:
```diff
@@ -4,3 +4,7 @@ import 'package:flutter/material.dart';
     GlobalKey<ScaffoldMessengerState>();
+
+/// refreshIndicator를 custom하게 변경하는 globalKey
+final GlobalKey<RefreshIndicatorState> refreshIndicatorKey =
```
이 파일에 있는 refreshIndicatorKey는 UserPage에서만 사용하고, 다른 페이지에서는 페이지 내부에서 _refreshIndicatorKey를 사용하는 방식으로 구현되어 있는데 사용방식의 통일이 필요할 것 같습니다.

sangohkim commented on lib/pages/user_page.dart:
```diff
@@ -361,79 +363,91 @@ class _UserPageState extends State<UserPage>
-                ),
-              ),
-            );
+    return NotificationListener<ScrollNotification>(
```
customRefreshIndicator가 이미 NotificationListener를 반환하는 형태로 구현되어 있어 수정이 필요할 것 같습니다. customRefreshIndicator의 적용방식이 UserPage에서만 조금씩다른 것 같은데(global key 사용 및 현재 부분에서) 혹시 다르게 구현하신 이유가 있을까요?

sangohkim reviewed (changes requested):
RefreshIndicator 직접 만드시느라 수고많으셨습니다! 리뷰 완료하였고 코멘트 달린 부분들 한번 봐주시면 될 것 같아요. 
그리고 한가지 수정해야할 부분이 보이는데 PostListShowPage, NotificationPage에서 pull down을 조금만 하면 의도했던 방식으로 리프레시가 진행되지만 pull down을 많이 하게되는 경우 페이지 중앙의 로딩인디케이터(저희가 저번에 페이지가 처음 로드될때만 보여주는걸로 한 부분)이 다시 나타나는 현상을 확인하였습니다. 저도 시간날때 원인을 파악해볼테니 이 부분도 확인부탁드립니다.
- iOS, android 모두에서 확인되었습니다

thomaskim1130 committed:
Fix: changed refresh dir

refresh_indicator.dart를 lib/widgets로 이동
user_page.dart에서 중복 implementation 제거

thomaskim1130 committed:
Fix: removed loadingindicator in some files

notificationPage, postListShowPage, postViewPage, userPage, userViewPage에서
isLoading 등 관련 변수들을 false 처리하면서 LoadingIndicator가 오직 처음에만 나타나도록 함.

userPage에서는 특히 TabType별로 global key를 분리하여 에러를 방지함.

thomaskim1130 commented:
Closed for now (outdated, continued in https://github.com/sparcs-kaist/new-ara-app/pull/246)
